pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,344 @@
+   *
+   * @param delayInMilliseconds pinot consumption delay being recorded.
+   * @param currentTime sample time.
+   * @param partitionId partition ID for which this delay is being recorded.
+   */
+  synchronized public void recordPinotConsumptionDelay(long delayInMilliseconds, long currentTime, long partitionId) {
+    if (delayInMilliseconds > _maxPinotConsumptionDelay) {
+      _maxPinotConsumptionDelay = delayInMilliseconds;
+      _maxPartitionId = partitionId;
+    }
+    // Store new measure and wipe old one for this partition
+    _partitionToDelaySampleMap.put(partitionId, new DelayMeasure(currentTime, delayInMilliseconds));
+    if (_partitionsMarkedForConfirmation.containsKey(partitionId)) {
+      // If we are consuming we do not need to track this partition for removal.
+      _partitionsMarkedForConfirmation.remove(partitionId);
+    }
+    if ((partitionId == _maxPartitionId) && (delayInMilliseconds < _maxPinotConsumptionDelay)) {
+      // If the partition that was maximum stops being maximum, we need to select a new maximum.
+      updateMaximum();
+    }
+  }
+
+  /*
+   * Handle partition removal event. This must be invoked when we stop serving a given partition for
+   * this table in the current server.
+   * This function will be invoked when we receive CONSUMING -> DROPPED / OFFLINE state transitions.
+   * Making it synchronized to support handling state changes concurrently.
+   *
+   * @param partitionId partition id that we should stop tracking.
+   */
+  synchronized public void stopTrackingPinotConsumptionDelayForPartition(long partitionId) {
+    removePartitionId(partitionId);
+  }
+
+  /*
+   * This method is used for ticking up the time a partition has been out of CONSUMING state.
+   * When that time exceeds some threshold, we read from ideal state to confirm we still host the partition, if
+   * not we remove the partition from being tracked locally.
+   * This call is to be invoked by a timer thread that will periodically wake up and invoke this function.
+   */
+  synchronized public void tickInactivePartitions() {
+    for (long partitionId : _partitionsMarkedForConfirmation.keySet()) {
+      int currentTicks = _partitionsMarkedForConfirmation.get(partitionId);
+      if (++currentTicks >= _maxNumberOfTicksBeforeVerification) {
+        if (!isPartitionHostedByThisServerPerIdealState(partitionId)) {
+          // If this partition is not hosted by this host as per ideal state we stop tracking
+          removePartitionId(partitionId);
+          return;
+        } else {
+          // Partition is hosted by us as per ideal state, hence do not track for removal
+          // Reset timeout and try later if it continues to not consume.
+          _partitionsMarkedForConfirmation.put(partitionId, 0);
+        }
+      } else {
+        // If not enough time in inactive state, we just add to the timeout count
+        _partitionsMarkedForConfirmation.put(partitionId, currentTicks);
+      }
```

Comment:
This is subjective:
Tick is another concept for the reader to understand. Instead we can keep the mapping of partitionId to the **time** that it's marked for IdealState verification. That way we don't need to increase ticks, we just compare the current time with the initial recording time.

Reply:
This was refactored as p[er our offline conversation